lib/ecto/adapters/sql.ex

```diff
@@ -91,6 +91,34 @@ defmodule Ecto.Adapters.SQL do
       end
 
       @doc false
+      def upsert(repo, %{source: {prefix, source}}, params, on_conflict, conflict_target, update, returning, opts) do
+        {fields, ins_values} = :lists.unzip(params)
+        values = case on_conflict do
+          :update ->
+            checked_update = Enum.filter(update, fn update_key -> Keyword.has_key?(params, update_key) end)
+            update_values = Enum.map(checked_update, fn update_key -> Keyword.get(params, update_key) end)
+            ins_values ++ update_values
+          _ -> ins_values
+        end
+        sql = @conn.upsert(prefix, source, fields, [fields], on_conflict, conflict_target, update, returning)
```

Comment:
If we're calling `upsert` on the connection, maybe we should also add `upsert` as a callback on the `Ecto.Adapters.SQL.Connection` behavior? Even if not a database doesn't supports upsert, they could still add a function to raise with an appropriate error message.